File: js/src/core/form-view.ts
import { BaseView } from './base-view';
import {
    // BaseViewConfig,
    FormFieldParams,
    SubmitAction,
    FieldValidation,
    ValidationResult,
    HttpMethod
} from '../types';
import { FieldValidator, FormValidator } from '../utils/validators';
import {
    MissingRequiredParameterError,
    FieldValidationError,
    FieldNotFoundError,
    EmptyCollectionError,
    Result,
    Ok,
    Err
} from '../errors';

export interface FormContent {
    title: string;
    intro?: string;
    submit?: SubmitAction;
    fields: Array<FormFieldParams & { fieldType: string; fieldId: string; fieldLabel: string }>;
}

/**
 * Builds a Form SGUI view — an interactive data-entry form.
 *
 * Fields are appended via `addField` and the typed helpers (`addTextField`,
 * `addEmailField`, `addSelectField`, `addPhotoField`, `addGPSField`, ...);
 * `submitButton`/`updateButton`/`deleteButton` define the submit action, and
 * `injectData`/`setFieldValue` pre-fill existing fields.
 *
 * Extends {@link BaseView}; instantiated by the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized to a JSON view description and
 * signed into a v3 envelope by `serve()`.
 */
export class FormView extends BaseView {

    static fromJson(json: Record<string, unknown>): FormView {
        return FormView.fromJsonAs(FormView, 'Form', json);
    }
    private fieldValidations: Map<string, FieldValidation> = new Map();

    constructor(formId: string, title: string, processId?: string) {
        super({
            id: formId,
            type: 'Form',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title,
            intro: '',
            submit: undefined,
            fields: []
        } as FormContent;
    }

    /**
     * Sets the form introduction (like ActionList/ActionGrid/etc.)
     */
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    /**
     * Helper method to associate this form with a process
     * @param processId - Process identifier
     * @param options - Process options (name, steps, etc.)
     */
    belongsToProcess(
        processId: string,
        options?: {
            processName?: string;
            currentStep?: number;
            totalSteps?: number;
            stepName?: string;
            canGoBack?: boolean;
            canSkip?: boolean;
        }
    ): this {
        this.setProcess(processId, options);
        return this;
    }

    /**
     * Adds a field with validation
     */
    addField(
        fieldType: string,
        fieldId: string,
        fieldLabel: string,
        params?: FormFieldParams
    ): this {
        if (!fieldId || !fieldLabel || !fieldType) {
            throw new MissingRequiredParameterError('fieldId, fieldLabel, and fieldType');
        }

        // Field validation
        const validation = FieldValidator.validateField(fieldType, fieldId, fieldLabel, params);
        if (!validation.isValid) {
            const errorMessages = validation.errors.map(e => e.message);
            throw new FieldValidationError(fieldId, fieldType, errorMessages);
        }

        const field = { fieldType, fieldId, fieldLabel, ...params };
        (this.content as FormContent).fields.push(field);

        // Store the validation for this field
        if (params) {
            this.fieldValidations.set(fieldId, params);
        }

        return this;
    }

    /**
     * Defines the submit button for the form
     * Convention: Mobile app POSTs to {service.baseUrl}/{formId}
     *
     * @param text - Button text (e.g., "Register", "Submit")
     * @param method - HTTP method (default: POST)
     * @param confirmMessage - Optional confirmation message for destructive actions
     */
    submitButton(text: string, method: HttpMethod = 'POST', confirmMessage?: string): this {
        (this.content as FormContent).submit = {
            text,
            method,
            confirmMessage
        };

        return this;
    }

    /**
     * Convenience method for update actions (PUT)
     */
    updateButton(text: string, confirmMessage?: string): this {
        return this.submitButton(text, 'PUT', confirmMessage);
    }

    /**
     * Convenience method for delete actions (DELETE with confirmation)
     */
    deleteButton(text: string, confirmMessage: string = 'Are you sure you want to delete this?'): this {
        return this.submitButton(text, 'DELETE', confirmMessage);
    }

    /**
     * Convenience methods for different field types
     */
    addTextField(fieldId: string, fieldLabel: string, isRequired: boolean = false, maxLength?: number): this {
        return this.addField('text', fieldId, fieldLabel, {
            required: isRequired,
            maxLength
        });
    }

    addEmailField(fieldId: string, fieldLabel: string, isRequired: boolean = false): this {
        return this.addField('email', fieldId, fieldLabel, {
            required: isRequired,
            pattern: /^[^\s@]+@[^\s@]+\.[^\s@]+$/
        });
    }

    addPasswordField(fieldId: string, fieldLabel: string, minLength: number = 8): this {
        return this.addField('password', fieldId, fieldLabel, {
            required: true,
            minLength
        });
    }

    addNumberField(fieldId: string, fieldLabel: string, isRequired: boolean = false, minVal?: number, maxVal?: number): this {
        return this.addField('number', fieldId, fieldLabel, {
            required: isRequired,
            min: minVal,
            max: maxVal
        });
    }

    addDateField(fieldId: string, fieldLabel: string, isRequired: boolean = false, minDate?: string, maxDate?: string): this {
        return this.addField('date', fieldId, fieldLabel, {
            required: isRequired,
            min: minDate ? new Date(minDate).getTime() : undefined,
            max: maxDate ? new Date(maxDate).getTime() : undefined
        });
    }

    addSelectField(fieldId: string, fieldLabel: string, isRequired: boolean = false, options: Array<{ label: string; value: unknown }>): this {
        if (!Array.isArray(options) || options.length === 0) {
            throw new EmptyCollectionError('Select field options', 'Select field must have at least one option');
        }

        return this.addField('select', fieldId, fieldLabel, {
            required: isRequired,
            options
        });
    }

    addPhotoField(fieldId: string, fieldLabel: string, isRequired: boolean = false, formats: string[] = ['jpeg', 'png'], live: boolean = false): this {
        if (!formats || formats.length === 0) {
            throw new EmptyCollectionError('Photo field formats', 'Photo field must specify at least one format');
        }

        const acceptedFormats = formats.map(format => `image/${format.toLowerCase()}`);
        return this.addField('photo', fieldId, fieldLabel, {
            required: isRequired,
            accept: acceptedFormats,
            live
        });
    }

    addFileField(fieldId: string, fieldLabel: string, isRequired: boolean = false, formats: string[]): this {
        if (!formats || formats.length === 0) {
            throw new EmptyCollectionError('File field formats', 'File field must specify at least one format');
        }

        return this.addField('file', fieldId, fieldLabel, {
            required: isRequired,
            accept: formats
        });
    }

    /**
     * @param config.altitude    include altitude in the captured value.
     * @param config.maxAccuracy the coarsest fix the provider accepts, in
     *   METRES (i.e. the minimum required precision). The mobile app keeps
     *   searching until `position.accuracy <= maxAccuracy`, and the captured
     *   accuracy travels back in the submitted value. Omit for the app default.
     * @param config.precision   legacy boolean high-accuracy flag — superseded
     *   by `maxAccuracy`; kept for backward compatibility.
     */
    addGPSField(
        fieldId: string,
        fieldLabel: string,
        isRequired: boolean = false,
        liveData: boolean = false,
        config: { altitude?: boolean; maxAccuracy?: number; precision?: boolean } = {}
    ): this {
        return this.addField('gps', fieldId, fieldLabel, {
            required: isRequired,
            live: liveData,
            ...config
        });
    }

    addPlusCodeField(fieldId: string, fieldLabel: string, isRequired: boolean = false, liveData: boolean = false): this {
        return this.addField('pluscode', fieldId, fieldLabel, {
            required: isRequired,
            live: liveData
        });
    }

    addHiddenField(fieldId: string, fieldLabel: string, value: string): this {
        return this.addField('hidden', fieldId, fieldLabel, { value });
    }

    addTextAreaField(fieldId: string, fieldLabel: string, isRequired: boolean = false, minLength?: number, maxLength?: number): this {
        return this.addField('textarea', fieldId, fieldLabel, {
            required: isRequired,
            minLength,
            maxLength
        });
    }

    addPhoneField(fieldId: string, fieldLabel: string, isRequired: boolean = false): this {
        return this.addField('phone', fieldId, fieldLabel, {
            required: isRequired,
            pattern: /^[\+]?[1-9][\d]{0,15}$/
        });
    }

    addURLField(fieldId: string, fieldLabel: string, isRequired: boolean = false): this {
        return this.addField('url', fieldId, fieldLabel, {
            required: isRequired
        });
    }

    addCheckboxField(fieldId: string, fieldLabel: string, isRequired: boolean = false): this {
        return this.addField('checkbox', fieldId, fieldLabel, {
            required: isRequired
        });
    }

    /**
     * Adds a visual separator to group form fields
     * Separators are rendered as gaps or lines by the mobile app renderer
     * @param fieldId - Optional field ID. If not provided, auto-generates a unique ID
     * @returns this for chaining
     * @example
     * form.addTextField('name', 'Name', true)
     *     .addSeparator()
     *     .addEmailField('email', 'Email', true);
     */
    addSeparator(fieldId?: string): this {
        const separatorId = fieldId || `separator-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
        return this.addField('separator', separatorId, '', {});
    }

    /**
     * Injects data into the form's existing fields
     * Automatically handles fields with options (select, radio) using "selected"
     * @param data - Key-value object where key = fieldId
     * @returns Result with errors for fields not found
     * @example
     * const result = form.injectData({
     *   name: 'John Doe',
     *   email: 'john@example.com',
     *   country: 'FR'  // For select: sets selected: true on the 'FR' option
     * });
     * if (result.isErr()) {
     *   console.warn('Some fields not found:', result.error);
     * }
     */
    injectData(data: Record<string, unknown>): Result<void, string[]> {
        const errors: string[] = [];

        Object.entries(data).forEach(([fieldId, value]) => {
            const field = this.getField(fieldId);
            if (!field) {
                errors.push(`Field not found: ${fieldId}`);
                return;
            }

            // Automatic detection: fields with options (select, radio, checkbox with options)
            if (field.options && field.fieldType === 'select') {
                // For select: mark the matching option as selected
                const updatedOptions = field.options.map(opt => ({
                    ...opt,
                    selected: opt.value === value
                }));
                this.updateField(fieldId, { options: updatedOptions as any });
            }
            // Radio: a single value
            else if (field.options && field.fieldType === 'radio') {
                const updatedOptions = field.options.map(opt => ({
                    ...opt,
                    selected: opt.value === value
                }));
                this.updateField(fieldId, { options: updatedOptions as any });
            }
            // Checkbox with options: value is an array (multi-select)
            else if (field.options && field.fieldType === 'checkbox') {
                const selectedValues = Array.isArray(value) ? value : [value];
                const updatedOptions = field.options.map(opt => ({
                    ...opt,
                    selected: selectedValues.includes(opt.value)
                }));
                this.updateField(fieldId, { options: updatedOptions as any });
            }
            // Simple checkbox (true/false) or all other fields: use value
            else {
                this.updateField(fieldId, { value: value as any });
            }
        });

        if (errors.length > 0) {
            return Err(errors);
        }
        return Ok(undefined);
    }

    /**
     * Sets the value of a specific field
     * @param fieldId - Field ID
     * @param value - Value to inject
     * @example
     * form.setFieldValue('name', 'John Doe')
     */
    setFieldValue(fieldId: string, value: unknown): this {
        const field = this.getField(fieldId);
        if (!field) {
            throw new FieldNotFoundError(fieldId, this.id);
        }

        this.updateField(fieldId, { value: value as any });
        return this;
    }

    /**
     * Validates the form data
     * Note: This is for SDK internal validation. Services using this SDK should
     * perform their own data validation before populating views.
     */
    validateFormData(formData: Record<string, any>): ValidationResult {
        return FormValidator.validateFormData(formData, this.fieldValidations);
    }

    /**
     * Gets a field by its ID
     */
    getField(fieldId: string): (FormFieldParams & { fieldType: string; fieldId: string; fieldLabel: string }) | undefined {
        return (this.content as FormContent).fields.find(field => field.fieldId === fieldId);
    }

    /**
     * Removes a field by its ID
     * @returns Result with success or error message
     */
    removeField(fieldId: string): Result<void, string> {
        const fields = (this.content as FormContent).fields;
        const index = fields.findIndex(field => field.fieldId === fieldId);

        if (index !== -1) {
            fields.splice(index, 1);
            this.fieldValidations.delete(fieldId);
            return Ok(undefined);
        }

        return Err(`Field '${fieldId}' not found in form '${this.id}'`);
    }

    /**
     * Updates an existing field
     * @returns Result with success or error message
     */
    updateField(fieldId: string, updates: Partial<FormFieldParams>): Result<void, string> {
        const field = this.getField(fieldId);
        if (!field) {
            return Err(`Field '${fieldId}' not found in form '${this.id}'`);
        }

        Object.assign(field, updates);

        // Update the validation
        if (this.fieldValidations.has(fieldId)) {
            const existingValidation = this.fieldValidations.get(fieldId)!;
            this.fieldValidations.set(fieldId, { ...existingValidation, ...updates });
        }

        return Ok(undefined);
    }

    /**
     * Gets all fields
     */
    getFields(): Array<FormFieldParams & { fieldType: string; fieldId: string; fieldLabel: string }> {
        return [...(this.content as FormContent).fields];
    }

    /**
     * Gets the field count
     * @param excludeSeparators - If true, excludes separator fields from count (default: false)
     */
    getFieldCount(excludeSeparators: boolean = false): number {
        if (excludeSeparators) {
            return (this.content as FormContent).fields.filter(field => field.fieldType !== 'separator').length;
        }
        return (this.content as FormContent).fields.length;
    }

    /**
     * Checks whether the form has required fields
     */
    hasRequiredFields(): boolean {
        return (this.content as FormContent).fields.some(field => field.required);
    }

    /**
     * Gets the required fields
     */
    getRequiredFields(): string[] {
        return (this.content as FormContent).fields
            .filter(field => field.required)
            .map(field => field.fieldId);
    }
}

File: py/yeriasdk/views/form_view.py
"""
FormView - A view for displaying forms with various field types
"""

from typing import Optional, Dict, Any, List
from datetime import datetime
import re

from ..core.base_view import BaseView
from ..types.models import FormFieldParams, SubmitAction, FieldValidation, ValidationResult, HttpMethod
from ..utils.validators import FieldValidator, FormValidator
from ..errors.exceptions import (
    MissingRequiredParameterError,
    FieldValidationError,
    FieldNotFoundError,
    EmptyCollectionError,
)


class FormView(BaseView):
    """View for displaying forms with validation"""

    def __init__(self, form_id: str, title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": form_id,
                "type": "Form",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": title,
            "intro": "",
            "submit": None,
            "fields": [],
        }

        self._field_validations: Dict[str, FieldValidation] = {}

    def set_intro(self, intro: str) -> "FormView":
        """Set form introduction text"""
        return self._set_intro_text("intro", intro)

    def set_note(self, note: str) -> "FormView":
        """Set form note (deprecated - use set_intro() instead)"""
        return self.set_intro(note)

    def belongs_to_process(
        self,
        process_id: str,
        process_name: Optional[str] = None,
        current_step: Optional[int] = None,
        total_steps: Optional[int] = None,
        step_name: Optional[str] = None,
        can_go_back: Optional[bool] = None,
        can_skip: Optional[bool] = None,
    ) -> "FormView":
        """Helper method to associate this form with a process"""
        context = {}
        if process_name:
            context["processName"] = process_name
        if current_step is not None:
            context["currentStep"] = current_step
        if total_steps is not None:
            context["totalSteps"] = total_steps
        if step_name:
            context["stepName"] = step_name
        if can_go_back is not None:
            context["canGoBack"] = can_go_back
        if can_skip is not None:
            context["canSkip"] = can_skip

        self.set_process(process_id, context)
        return self

    def add_field(
        self,
        field_type: str,
        field_id: str,
        field_label: str,
        params: Optional[FormFieldParams] = None,
    ) -> "FormView":
        """Add a field with validation"""
        if not field_id or not field_label or not field_type:
            raise MissingRequiredParameterError("fieldId, fieldLabel, and fieldType")

        # Validate the field
        validation = FieldValidator.validate_field(
            field_type, field_id, field_label, params
        )
        if not validation.is_valid:
            error_messages = [e.message for e in validation.errors]
            raise FieldValidationError(field_id, field_type, error_messages)

        field = {
            "fieldType": field_type,
            "fieldId": field_id,
            "fieldLabel": field_label,
        }

        if params:
            # Add all params to field
            if params.value is not None:
                field["value"] = params.value
            if params.required is not None:
                field["required"] = params.required
            if params.pattern:
                # Convert regex pattern to string representation
                field["pattern"] = params.pattern.pattern
            if params.min is not None:
                field["min"] = params.min
            if params.max is not None:
                field["max"] = params.max
            if params.min_length is not None:
                field["minLength"] = params.min_length
            if params.max_length is not None:
                field["maxLength"] = params.max_length
            if params.options:
                field["options"] = [
                    {
                        "label": opt.label,
                        "value": opt.value,
                        "selected": opt.selected,
                    }
                    for opt in params.options
                ]
            if params.accept:
                field["accept"] = params.accept
            if params.live is not None:
                field["live"] = params.live
            if params.placeholder:
                field["placeholder"] = params.placeholder
            if params.help_text:
                field["helpText"] = params.help_text
            if params.disabled is not None:
                field["disabled"] = params.disabled
            if params.readonly is not None:
                field["readonly"] = params.readonly
            if params.min_date:
                field["minDate"] = params.min_date
            if params.max_date:
                field["maxDate"] = params.max_date

            # Store validation for this field
            self._field_validations[field_id] = params

        self.content["fields"].append(field)
        return self

    def submit_button(
        self, text: str, method: HttpMethod = "POST", confirm_message: Optional[str] = None
    ) -> "FormView":
        """Define submit button for the form"""
        self.content["submit"] = {
            "text": text,
            "method": method,
            "confirmMessage": confirm_message,
        }
        return self

    def update_button(
        self, text: str, confirm_message: Optional[str] = None
    ) -> "FormView":
        """Convenience method for update actions (PUT)"""
        return self.submit_button(text, "PUT", confirm_message)

    def delete_button(
        self, text: str, confirm_message: str = "Are you sure you want to delete this?"
    ) -> "FormView":
        """Convenience method for delete actions (DELETE with confirmation)"""
        return self.submit_button(text, "DELETE", confirm_message)

    # Convenience methods for different field types
    def add_text_field(
        self, field_id: str, field_label: str, is_required: bool = False, max_length: Optional[int] = None
    ) -> "FormView":
        params = FormFieldParams(required=is_required, max_length=max_length)
        return self.add_field("text", field_id, field_label, params)

    def add_email_field(
        self, field_id: str, field_label: str, is_required: bool = False
    ) -> "FormView":
        params = FormFieldParams(required=is_required, pattern=re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$"))
        return self.add_field("email", field_id, field_label, params)

    def add_password_field(
        self, field_id: str, field_label: str, min_length: int = 8
    ) -> "FormView":
        params = FormFieldParams(required=True, min_length=min_length)
        return self.add_field("password", field_id, field_label, params)

    def add_number_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        min_val: Optional[float] = None,
        max_val: Optional[float] = None,
    ) -> "FormView":
        params = FormFieldParams(required=is_required, min=min_val, max=max_val)
        return self.add_field("number", field_id, field_label, params)

    def add_date_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        min_date: Optional[str] = None,
        max_date: Optional[str] = None,
    ) -> "FormView":
        params = FormFieldParams(required=is_required, min_date=min_date, max_date=max_date)
        return self.add_field("date", field_id, field_label, params)

    def add_select_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        options: Optional[List[Dict[str, Any]]] = None,
    ) -> "FormView":
        if not options or len(options) == 0:
            raise EmptyCollectionError(
                "Select field options", "Select field must have at least one option"
            )

        from ..types.models import FormFieldOption
        field_options = [
            FormFieldOption(
                label=opt["label"],
                value=opt["value"],
                selected=opt.get("selected", False),
            )
            for opt in options
        ]

        params = FormFieldParams(required=is_required, options=field_options)
        return self.add_field("select", field_id, field_label, params)

    def add_photo_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        formats: Optional[List[str]] = None,
        live: bool = False,
    ) -> "FormView":
        if not formats:
            formats = ["jpeg", "png"]
        if len(formats) == 0:
            raise EmptyCollectionError(
                "Photo field formats", "Photo field must specify at least one format"
            )

        accepted_formats = [f"image/{f.lower()}" for f in formats]
        params = FormFieldParams(required=is_required, accept=accepted_formats, live=live)
        return self.add_field("photo", field_id, field_label, params)

    def add_file_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        formats: Optional[List[str]] = None,
    ) -> "FormView":
        if not formats or len(formats) == 0:
            raise EmptyCollectionError(
                "File field formats", "File field must specify at least one format"
            )

        params = FormFieldParams(required=is_required, accept=formats)
        return self.add_field("file", field_id, field_label, params)

    def add_gps_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        live_data: bool = False,
        altitude: Optional[bool] = None,
        precision: Optional[bool] = None,
    ) -> "FormView":
        params = FormFieldParams(required=is_required, live=live_data)
        return self.add_field("gps", field_id, field_label, params)

    def add_plus_code_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        live_data: bool = False,
    ) -> "FormView":
        params = FormFieldParams(required=is_required, live=live_data)
        return self.add_field("pluscode", field_id, field_label, params)

    def add_hidden_field(
        self, field_id: str, field_label: str, value: str
    ) -> "FormView":
        params = FormFieldParams(value=value)
        return self.add_field("hidden", field_id, field_label, params)

    def add_text_area_field(
        self,
        field_id: str,
        field_label: str,
        is_required: bool = False,
        min_length: Optional[int] = None,
        max_length: Optional[int] = None,
    ) -> "FormView":
        params = FormFieldParams(
            required=is_required, min_length=min_length, max_length=max_length
        )
        return self.add_field("textarea", field_id, field_label, params)

    def add_phone_field(
        self, field_id: str, field_label: str, is_required: bool = False
    ) -> "FormView":
        params = FormFieldParams(required=is_required, pattern=re.compile(r"^[\+]?[1-9][\d]{0,15}$"))
        return self.add_field("phone", field_id, field_label, params)

    def add_url_field(
        self, field_id: str, field_label: str, is_required: bool = False
    ) -> "FormView":
        params = FormFieldParams(required=is_required)
        return self.add_field("url", field_id, field_label, params)

    def add_checkbox_field(
        self, field_id: str, field_label: str, is_required: bool = False
    ) -> "FormView":
        params = FormFieldParams(required=is_required)
        return self.add_field("checkbox", field_id, field_label, params)

    def add_separator(self, field_id: Optional[str] = None) -> "FormView":
        """Add a visual separator to group form fields

        Separators are rendered as gaps or lines by the mobile app renderer.

        Args:
            field_id: Optional field ID. If not provided, auto-generates a unique ID.

        Returns:
            self for chaining

        Example:
            form.add_text_field('name', 'Name', True)
                .add_separator()
                .add_email_field('email', 'Email', True)
        """
        import time
        import random
        separator_id = field_id or f"separator-{int(time.time() * 1000)}-{random.randint(1000, 9999)}"
        return self.add_field("separator", separator_id, "", None)

    def inject_data(self, data: Dict[str, Any]) -> List[str]:
        """Inject data into existing form fields"""
        errors: List[str] = []

        for field_id, value in data.items():
            field = self.get_field(field_id)
            if not field:
                errors.append(f"Field not found: {field_id}")
                continue

            # Auto-detect: fields with options (select, radio, checkbox with options)
            if field.get("options") and field.get("fieldType") == "select":
                updated_options = [
                    {**opt, "selected": opt["value"] == value}
                    for opt in field["options"]
                ]
                self.update_field(field_id, {"options": updated_options})
            elif field.get("options") and field.get("fieldType") == "radio":
                updated_options = [
                    {**opt, "selected": opt["value"] == value}
                    for opt in field["options"]
                ]
                self.update_field(field_id, {"options": updated_options})
            elif field.get("options") and field.get("fieldType") == "checkbox":
                selected_values = value if isinstance(value, list) else [value]
                updated_options = [
                    {**opt, "selected": opt["value"] in selected_values}
                    for opt in field["options"]
                ]
                self.update_field(field_id, {"options": updated_options})
            else:
                self.update_field(field_id, {"value": value})

        return errors

    def set_field_value(self, field_id: str, value: Any) -> "FormView":
        """Set the value of a specific field"""
        field = self.get_field(field_id)
        if not field:
            raise FieldNotFoundError(field_id, self.id)

        self.update_field(field_id, {"value": value})
        return self

    def validate_form_data(self, form_data: Dict[str, Any]) -> ValidationResult:
        """Validate form data"""
        return FormValidator.validate_form_data(form_data, self._field_validations)

    def get_field(self, field_id: str) -> Optional[Dict[str, Any]]:
        """Get a field by its ID"""
        return next(
            (f for f in self.content["fields"] if f.get("fieldId") == field_id), None
        )

    def remove_field(self, field_id: str) -> bool:
        """Remove a field by its ID"""
        fields = self.content["fields"]
        index = next(
            (i for i, f in enumerate(fields) if f.get("fieldId") == field_id), -1
        )

        if index != -1:
            fields.pop(index)
            if field_id in self._field_validations:
                del self._field_validations[field_id]
            return True

        return False

    def update_field(self, field_id: str, updates: Dict[str, Any]) -> bool:
        """Update an existing field"""
        field = self.get_field(field_id)
        if not field:
            return False

        field.update(updates)

        # Update validation
        if field_id in self._field_validations:
            existing_validation = self._field_validations[field_id]
            # Merge updates into validation
            for key, value in updates.items():
                if hasattr(existing_validation, key):
                    setattr(existing_validation, key, value)

        return True

    def get_fields(self) -> List[Dict[str, Any]]:
        """Get all fields"""
        return list(self.content["fields"])

    def get_field_count(self, exclude_separators: bool = False) -> int:
        """Get field count

        Args:
            exclude_separators: If True, excludes separator fields from count (default: False)

        Returns:
            Number of fields
        """
        if exclude_separators:
            return len([
                f for f in self.content["fields"]
                if f.get("fieldType") != "separator"
            ])
        return len(self.content["fields"])

    def has_required_fields(self) -> bool:
        """Check if form has required fields"""
        return any(f.get("required") for f in self.content["fields"])

    def get_required_fields(self) -> List[str]:
        """Get required field IDs"""
        return [
            f["fieldId"] for f in self.content["fields"] if f.get("required")
        ]

